apprise/Apprise.py

```diff
@@ -370,41 +368,15 @@ def notify(self, body, title='', notify_type=common.NotifyType.INFO,
         such as turning a \n into an actual new line, etc.
         """
 
-        if ASYNCIO_SUPPORT:
-            return py3compat.asyncio.tosync(
-                self.async_notify(
-                    body, title,
-                    notify_type=notify_type, body_format=body_format,
-                    tag=tag, match_always=match_always, attach=attach,
-                    interpret_escapes=interpret_escapes,
-                ),
-                debug=self.debug
-            )
-
-        else:
-            try:
-                results = list(
-                    self._notifyall(
-                        Apprise._notifyhandler,
-                        body, title,
-                        notify_type=notify_type, body_format=body_format,
-                        tag=tag, attach=attach,
-                        interpret_escapes=interpret_escapes,
-                    )
-                )
-
-            except TypeError:
-                # No notifications sent, and there was an internal error.
-                return False
-
-            else:
-                if len(results) > 0:
-                    # All notifications sent, return False if any failed.
-                    return all(results)
-
-                else:
-                    # No notifications sent.
-                    return None
+        return py3compat.asyncio.tosync(
+            self.async_notify(
+                body, title,
+                notify_type=notify_type, body_format=body_format,
+                tag=tag, match_always=match_always, attach=attach,
+                interpret_escapes=interpret_escapes,
+            ),
+            debug=self.debug
+        )
 
     def async_notify(self, *args, **kwargs):
```

Comment:
The docstring of this method says:

> This method is an async method that should be awaited on, even if it is missing the `async` keyword in its signature. This is omitted to preserve syntax compatibility with Python 2.

After dropping support for Python 2, the `async` keyword may be added back, if that makes sense?

Reply:
Yes, you can add `async`; I wrote the async_notify() method with just this scenario in mind. 🙂

Reply:
Let's not get hasty on this one... a lot of people have adopted `notify` into their code.   they'll all break and have to add `await` in front of it and start tracking loops.  Leave the `async_notify()` as is for now.

Reply:
No need to be worried; there is no backwards compatibility issue. `notify()` will continue to exist, and you already had to write ` await async_notify()`. Switching to
```python
def async_notify(self, *args, **kwargs)
```
just makes it a little more obvious that it must be called with `await`. 😉

Reply:
Indeed, if you look at the [corresponding .pyi file](https://github.com/caronc/apprise/blob/f7244cce3da54c5a3dc06a9603faa0e3cc6a47b0/apprise/Apprise.pyi#L46), you'll see that the type annotation already has the `async` keyword included.

Reply:
Hi Ryan, nice to meet you. I think that Chris has a point here. Even if that function has been _declared_ `async`, it has not been _defined_ as such. So, chances are, that people are using it within synchronous code. Even if it would have been wrong not to await the future, their code would still have mostly worked. I think we should be polite enough to add a runtime warning now, which will announce the switchover to `async` will happen with the next major release, Apprise 2.0.0. If you don't have any objections, I will add such a warning.

Reply:
I think what he was pointing out was just that we can stick the keyword `async` now on the `async_notify()` function.  At this time it isn't' present.  I got confused too and thought he was targeting the `notify()` call; but that's not the case :+1: 

I can actually do this now while I'm here in the code base just to see how it handles.

Reply:
Also pleased to meet you, Andreas!

> Even if that function has been _declared_ `async`, it has not been _defined_ as such. So, chances are, that people are using it within synchronous code. Even if it would have been wrong not to await the future, their code would still have mostly worked.

The function already returns a future, hence every return value being wrapped with `py3compat.asyncio.toasyncwrapvalue`. I do believe it's indistinguishable from an `async` function with the wrapper removed. Because if you don't use `await`, you have to unwrap the future.

Reply:
> After touching some stuff here and blowing it all up, i think I'll leave it the way it is for this commit since everything else works so well. @YoRyan is an async master... he breaths in air and outputs asynchronous threaded Python 😉. i think i'll wait for him to review the `async` part when this request is all pushed upstream.

Aww, thanks. 😁 Yes, we absolutely can leave this for later, and I would be happy to make that PR.

For the record, if you add the `async` keyword, it's also necessary to remove the `py3compat.asyncio.toasyncwrapvalue` calls. Basically, the following two functions are equivalent:

```python
def myAsyncFunction():
    return py3compat.asyncio.toasyncwrapvalue("Hello, World!")
```
```python
async def myAsyncFunction():
    return "Hello, World!"
```

Reply:
Thanks for providing an example. Let me demonstrate, while the _returning behavior_ is being made equivalent, the _invocation behavior_, at runtime, is not.

The following code, where the `async_foo` function is called synchronously, will run all its main body, before issuing a `RuntimeWarning: coroutine 'toasyncwrapvalue' was never awaited` when _returning_.

On the other hand, when you will prefix that routine with the `async` keyword, the function _will not be invoked at all_. Instead, Python will issue a `RuntimeWarning: coroutine 'foo_async' was never awaited`.

On that matter, I think Chris is absolutely right:

> A lot of people have adopted notify into their code. They'll all break and have to add `await` in front of it and start tracking loops. Leave the `async_notify()` as is for now.

However, it will not break because of not waiting for the completion of the function's future, that would only be about getting a return value anyway, and making sure it completed. It will break for all people who either unaware or errorneously are invoking the function _synchronously_, and then suddenly finding it would stop being executed, which might be _many_.

All users who are doing it this way, _must_ change their code to `await` the function, _and_ make sure they will invoke their code asynchronously with `asyncio.run(main())`. Many may not have heard about all of that at all.

I think this would be a change which should only be done on a major release bump, designated as "breaking change".

```
import apprise


def async_foo():
    raise ValueError("foo")
    return apprise.py3compat.asyncio.toasyncwrapvalue("Hello, World!")


def main():
    async_foo()


if __name__ == "__main__":
    main()
```

Reply:
Oops, I understand now. I had not considered synchronous callers who would not be interested in the return value and who might inadvertently discard the future.

I absolutely intended for async_notify() to behave just like an `async def` function, and it's unfortunate - given the constraint of having to be syntactically compatible with Python 2 - that I was unable to achieve a perfect replacement.

But in contrast to "a lot of people," I doubt this breakage would affect that many at all. The new async method was only introduced [fairly recently](https://github.com/caronc/apprise/pull/397), and the "async_" prefix combined with the type annotation _should_ have served as reminders that the returned future needs to be await'ed. Nevertheless, yes, it is a breakage, and I agree it should be documented for a future PR and release.

Reply:
Hi again,

> The new async method was only introduced _recently_.

It is still 1.5 years, though ;].

> But in contrast to "a lot of people," I doubt this breakage would affect that many at all. 

Sure, I think I might have exeggarated this a bit. I only wanted to outline the potential breaking change with insistence, but not overshoot the discussion. It may have happened - apologies for that.

> Nevertheless, yes, it is a breakage, and I agree it should be documented for a future PR and release.

I think Chris will be happy if you continue throwing in your expertise on this matter. As we observed yesterday with a testing build on Copr, just slapping an `async` keyword will not immediately work, but instead break some of Apprise's own tests.

Whether the improvement may be released within the 1.x series, with a corresponding admonition in the changelog, or just right away with a new 2.x release, is of course up to you. My verdict is: Why not just make a 2.x when the switchover is ready - I think it will not do any harm whatsoever.

Keep up the spirit and with kind regards,
Andreas.